### PR TITLE
Socket authentication hook

### PR DESCRIPTION
**Pull request by cschofieldffx**
A small update on the existing PR raised by fproulx-eoscanada

### CODE REVIEW

File: graphqlws/http.go [hunk 1]
```diff
@@ -16,7 +15,7 @@ var upgrader = websocket.Upgrader{
 }
 
 // NewHandlerFunc returns an http.HandlerFunc that supports GraphQL over websockets
-func NewHandlerFunc(svc connection.GraphQLService, httpHandler http.Handler) http.HandlerFunc {
+func NewHandlerFunc(svc connection.GraphQLService, httpHandler http.Handler, authFunc connection.AuthenticateFunc) http.HandlerFunc {
```
Comment (dackroyd) [inline graphqlws/http.go]: This is a breaking change to the `NewHandlerFunc`. The 'functional options' approach could be used to introduce this new parameter in a non-breaking way?

File: graphqlws/http.go [hunk 2]
```diff
@@ -4,7 +4,6 @@ import (
 	"net/http"
 
 	"github.com/gorilla/websocket"
-
```
Comment (dackroyd) [inline graphqlws/http.go]: Nit: unrelated

File: graphqlws/internal/connection/connection.go [hunk 3]
```diff
@@ -47,18 +48,28 @@ type startMessagePayload struct {
...
+
 type connection struct {
-	cancel       func()
-	service      GraphQLService
-	writeTimeout time.Duration
-	ws           wsConnection
+	authenticated    bool
+	authenticateFunc AuthenticateFunc
+	cancel           func()
+	request          *http.Request
```
Comment (dackroyd) [inline graphqlws/internal/connection/connection.go]: Is it ok to hold onto the original request here, or is that going to result in a leak? At a minimum, it may be heavy to hold onto the entire request for the lifetime of the connection?

File: graphqlws/http.go (hunk 1 above)
Comment (cschofieldffx) [reply graphqlws/http.go]: Can do.

File: graphqlws/internal/connection/connection.go (hunk 3 above)
Comment (cschofieldffx) [reply graphqlws/internal/connection/connection.go]: Probably don't need to hold on to the full request but the keeping the request headers would be very useful

File: graphqlws/http.go (hunk 1 above)
Comment (marcocali) [reply graphqlws/http.go]: There is a reference in the README.md to the original _NewHandlerFunc_ func.

File: graphqlws/internal/connection/connection.go [hunk 4]
```diff
@@ -159,15 +170,29 @@ func (conn *connection) readLoop(ctx context.Context, send sendFunc) {
...
-			var initMsg initMessagePayload
+			var initMsg map[string]interface{}
 			if err := json.Unmarshal(msg.Payload, &initMsg); err != nil {
 				ep := errPayload(fmt.Errorf("invalid payload for type: %s", msg.Type))
 				send("", typeConnectionError, ep)
 				continue
 			}
+
+			if conn.authenticateFunc != nil {
+				if err := conn.authenticateFunc(ctx, conn.request, initMsg); err != nil {
```
Comment (marcocali) [inline graphqlws/internal/connection/connection.go]: It seems like `initMsg` is only used here. Should the `json.Unmarshal` happen within this if block?

Comment (cschofieldffx) [reply graphqlws/internal/connection/connection.go]: Good point

File: graphqlws/http.go (hunk 1 above)
Comment (cschofieldffx) [reply graphqlws/http.go]: Will fix that to
